test/e2e/storage/vsphere_volume_perf.go

```diff
@@ -0,0 +1,264 @@
+
+	var (
+		client           clientset.Interface
+		namespace        string
+		nodeSelectorList []*NodeSelector
+		volumeCount      int
+		volumesPerPod    int
+		iterations       int
+	)
+
+	BeforeEach(func() {
+		var err error
+		framework.SkipUnlessProviderIs("vsphere")
+		client = f.ClientSet
+		namespace = f.Namespace.Name
+
+		Expect(os.Getenv("VSPHERE_SPBM_GOLD_POLICY")).NotTo(BeEmpty(), "env VSPHERE_SPBM_GOLD_POLICY is not set")
+		Expect(os.Getenv("VSPHERE_DATASTORE")).NotTo(BeEmpty(), "env VSPHERE_DATASTORE is not set")
+		Expect(os.Getenv("VCP_PERF_VOLUME_PER_POD")).NotTo(BeEmpty(), "env VCP_PERF_VOLUME_PER_POD is not set")
+		Expect(os.Getenv("VCP_PERF_ITERATIONS")).NotTo(BeEmpty(), "env VCP_PERF_ITERATIONS is not set")
+		Expect(os.Getenv("VCP_PERF_VOLUME_COUNT")).NotTo(BeEmpty(), "env VCP_PERF_VOLUME_COUNT is not set")
+
+		// Verify volume count specified by the user can be satisfied
+		volumeCount, err = strconv.Atoi(os.Getenv("VCP_PERF_VOLUME_COUNT"))
+		Expect(err).NotTo(HaveOccurred(), "Error Parsing VCP_PERF_VOLUME_COUNT")
+		volumesPerPod, err = strconv.Atoi(os.Getenv("VCP_PERF_VOLUME_PER_POD"))
+		Expect(err).NotTo(HaveOccurred(), "Error Parsing VCP_PERF_VOLUME_PER_POD")
+		iterations, err = strconv.Atoi(os.Getenv("VCP_PERF_ITERATIONS"))
+		Expect(err).NotTo(HaveOccurred(), "Error Parsing VCP_PERF_ITERATIONS")
+
+		nodes := framework.GetReadySchedulableNodesOrDie(client)
+		if len(nodes.Items) < 2 {
+			framework.Skipf("Requires at least %d nodes (not %d)", 2, len(nodes.Items))
+		}
+		if volumeCount > SCSIUnitsAvailablePerNode*len(nodes.Items) {
+			framework.Skipf("Cannot attach %d volumes to %d nodes. Maximum volumes that can be attached on %d nodes is %d", volumeCount, len(nodes.Items), len(nodes.Items), SCSIUnitsAvailablePerNode*len(nodes.Items))
+		}
+		nodeSelectorList = createNodeLabels(client, namespace, nodes)
+
+	})
+
+	It("vcp performance tests", func() {
+		scList := getTestStorageClasses(client)
+
+		var sumLatency [NumOps]int64
+		for i := 0; i < iterations; i++ {
+			latency := invokeVolumeLifeCyclePerformance(f, client, namespace, scList, volumesPerPod, volumeCount, nodeSelectorList)
+			for i, val := range latency {
+				sumLatency[i] += val
+			}
+		}
+
+		iterations64 := int64(iterations)
+		framework.Logf("Average latency for below operations")
+		framework.Logf("Creating %v PVCs and waiting for bound phase: %v seconds", volumeCount, sumLatency[0]/iterations64)
+		framework.Logf("Creating %v Pod: %v seconds", volumeCount/volumesPerPod, sumLatency[1]/iterations64)
+		framework.Logf("Deleting %v Pod and waiting for disk to be detached: %v seconds", volumeCount/volumesPerPod, sumLatency[2]/iterations64)
+		framework.Logf("Deleting %v PVCs: %v seconds", volumeCount, sumLatency[3]/iterations64)
+
+		for _, sc := range scList {
+			client.StorageV1().StorageClasses().Delete(sc.Name, nil)
+		}
+	})
+})
+
+func getTestStorageClasses(client clientset.Interface) []*storageV1.StorageClass {
```

Comment:
I didn't check, but isn't this code or very similar repeated in another vsphere test suite?

Reply:
Yes. 
As part of larger refactoring after all tests are merged, all the common code is supposed to be refactored as per vmware#353